test/utils/Checkpoints.t.sol

```diff
@@ -0,0 +1,276 @@
+        vm.assume(keys.length > 0);
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
+
+        uint32 lastKey = keys[keys.length - 1];
+        vm.assume(lastKey > 0);
+        lookup = _boundUint32(lookup, 0, lastKey - 1);
+
+        uint224 upper = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint32 key = keys[i];
+            uint224 value = values[i % values.length];
+
+            // push
+            vm.roll(key);
+            _ckpts.push(value);
+
+            // track expected result of lookups
+            if (key <= lookup) {
+                upper = value;
+            }
+        }
+
+        // check lookup
+        assertTrue(_ckpts.getAtBlock(lookup) == upper);
+        assertTrue(_ckpts.getAtProbablyRecentBlock(lookup) == upper);
+
+        vm.expectRevert();
+        _ckpts.getAtBlock(lastKey);
+        vm.expectRevert();
+        _ckpts.getAtBlock(lastKey + 1);
+        vm.expectRevert();
+        _ckpts.getAtProbablyRecentBlock(lastKey);
+        vm.expectRevert();
+        _ckpts.getAtProbablyRecentBlock(lastKey + 1);
+    }
+}
+
+contract CheckpointsTrace224Test is Test {
+    using Checkpoints for Checkpoints.Trace224;
+
+    Checkpoints.Trace224 internal _ckpts;
+
+    // helpers
+    function _boundUint32(uint32 x, uint32 min, uint32 max) internal view returns (uint32) {
+        return SafeCast.toUint32(bound(uint256(x), uint256(min), uint256(max)));
+    }
+
+    function _prepareKeys(uint32[] memory keys, uint32 maxSpread) internal view {
+        uint32 lastKey = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint32 key = _boundUint32(keys[i], lastKey, lastKey + maxSpread);
+            keys[i] = key;
+            lastKey = key;
+        }
+    }
+
+    function _assertLatestCheckpoint(bool exist, uint32 key, uint224 value) internal {
+        (bool _exist, uint32 _key, uint224 _value) = _ckpts.latestCheckpoint();
+        assertTrue(_exist == exist);
+        assertTrue(_key == key);
+        assertTrue(_value == value);
+    }
+
+    // tests
+    function testPush(uint32[] memory keys, uint224[] memory values) public {
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
+
+        // initial state
+        assertTrue(_ckpts.length() == 0);
+        assertTrue(_ckpts.latest() == 0);
+        _assertLatestCheckpoint(false, 0, 0);
+
+        uint256 duplicates = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint32 key = keys[i];
+            uint224 value = values[i % values.length];
+            if (i > 0 && key == keys[i - 1]) ++duplicates;
+
+            // push
+            _ckpts.push(key, value);
+
+            // check length & latest
+            assertTrue(_ckpts.length() == i + 1 - duplicates);
+            assertTrue(_ckpts.latest() == value);
+            _assertLatestCheckpoint(true, key, value);
+        }
```

Comment:
Perhaps we can add a final push after the loop, with a random key lower than the latest checkpoint, and check that it reverts?